x-pack/plugins/security_solution/public/common/components/exceptions/viewer/index.tsx

```diff
@@ -4,96 +4,412 @@
-import { ExceptionEntries } from './exception_entries';
-import { getFormattedEntries, getFormattedComments } from '../helpers';
-import { FormattedEntry, ExceptionListItemSchema } from '../types';
+import * as i18n from '../translations';
+import { useStateToaster } from '../../toasters';
+import { useKibana } from '../../../../common/lib/kibana';
+import { Panel } from '../../../../common/components/panel';
+import { Loader } from '../../../../common/components/loader';
+import { ExceptionsViewerHeader } from './exceptions_viewer_header';
+import {
+  ExceptionListType,
+  ExceptionListItemSchema,
+  ApiProps,
+  Filter,
+  SetExceptionsProps,
+} from '../types';
+import { allExceptionItemsReducer, State } from './reducer';
+import {
+  useExceptionList,
+  ExceptionIdentifiers,
+  useApi,
+} from '../../../../../public/lists_plugin_deps';
+import { ExceptionItem } from './exception_item';
+import { AndOrBadge } from '../../and_or_badge';
+import { ExceptionsViewerPagination } from './exceptions_pagination';
+import {
+  UtilityBar,
+  UtilityBarSection,
+  UtilityBarGroup,
+  UtilityBarText,
+  UtilityBarAction,
+} from '../../utility_bar';
 
-const MyFlexItem = styled(EuiFlexItem)`
-    &.comments--show {
-      padding: ${({ theme }) => theme.eui.euiSize};
-      border-top: ${({ theme }) => `${theme.eui.euiBorderThin}`}
+const StyledText = styled(EuiText)`
+  font-style: italic;
+`;
 
+const MyExceptionsContainer = styled.div`
+  height: 600px;
+  overflow: hidden;
 `;
 
-interface ExceptionItemProps {
-  exceptionItem: ExceptionListItemSchema;
+const initialState: State = {
+  filterOptions: { filter: '', showEndpointList: false, showDetectionsList: false, tags: [] },
+  pagination: {
+    pageIndex: 0,
+    pageSize: 20,
+    totalItemCount: 0,
+    pageSizeOptions: [5, 10, 20, 50, 100, 200, 300],
+  },
+  endpointList: null,
+  detectionsList: null,
+  allExceptions: [],
+  exceptions: [],
+  exceptionToEdit: null,
+  loadingItemIds: [],
+  isModalOpen: false,
+};
+
+enum ModalAction {
+  CREATE = 'CREATE',
+  EDIT = 'EDIT',
+}
+
+interface ExceptionsViewerProps {
+  ruleId: string;
+  exceptionListsMeta: ExceptionIdentifiers[];
+  availableListTypes: ExceptionListType[];
   commentsAccordionId: string;
-  handleDelete: ({ id }: { id: string }) => void;
-  handleEdit: (item: ExceptionListItemSchema) => void;
+  onAssociateList?: (listId: string) => void;
 }
 
-const ExceptionItemComponent = ({
-  exceptionItem,
+const ExceptionsViewerComponent = ({
+  ruleId,
+  exceptionListsMeta,
+  availableListTypes,
+  onAssociateList,
   commentsAccordionId,
-  handleDelete,
-  handleEdit,
-}: ExceptionItemProps): JSX.Element => {
-  const [entryItems, setEntryItems] = useState<FormattedEntry[]>([]);
-  const [showComments, setShowComments] = useState(false);
+}: ExceptionsViewerProps): JSX.Element => {
+  const { services } = useKibana();
+  const [, dispatchToaster] = useStateToaster();
+  const [initLoading, setInitLoading] = useState(true);
```

Comment:
Since you are keeping track of most of the state in a reducer, is it possible to also keep track of this piece of state in there?